pkg/util/commands.go

```diff
@@ -1,96 +1,164 @@
+}
+
+// Attempts The number of times the command has been executed
+func (c *Command) Attempts() int {
+	return c.attempts
+}
+
+// DidError returns a boolean if any error occurred in any execution of the command
+func (c *Command) DidError() bool {
+	if len(c.Errors) > 0 {
+		return true
 	}
-	return answer
+	return false
 }
 
-// RunCommandWithOutput evaluates the given command and returns the trimmed output
-func RunCommandWithOutput(dir string, name string, args ...string) (string, error) {
-	os.Setenv("PATH", PathWithBinary())
-	e := exec.Command(name, args...)
-	if dir != "" {
-		e.Dir = dir
+// DidFail returns a boolean if the command could not complete (errored on every attempt)
+func (c *Command) DidFail() bool {
+	if len(c.Errors) == c.attempts {
+		return true
 	}
-	data, err := e.CombinedOutput()
-	text := string(data)
-	text = strings.TrimSpace(text)
-	if err != nil {
-		return text, errors.Wrapf(err, "failed to run '%s %s' command in directory '%s', output: '%s'",
-			name, strings.Join(args, " "), dir, text)
+	return false
+}
+
+// Error returns the last error
+func (c *Command) Error() error {
+	if len(c.Errors) > 0 {
+		return c.Errors[len(c.Errors)-1]
 	}
-	return text, err
+	return nil
+}
+
+// IsVerbose Evaluates options and returns if Command is set to verbose
+func (c *Command) IsVerbose() bool {
+	return c.Verbose
 }
 
-// RunCommand evaluates the given command and returns the trimmed output
-func RunCommand(dir string, name string, args ...string) error {
-	os.Setenv("PATH", PathWithBinary())
-	e := exec.Command(name, args...)
-	if dir != "" {
-		e.Dir = dir
+// IsQuiet Evaluates options and returns if Command is set to quiet
+func (c *Command) IsQuiet() bool {
+	if c.IsVerbose() {
+		return false
 	}
-	var b bytes.Buffer
-	e.Stdout = &b
-	e.Stderr = &b
-	err := e.Run()
-	output := string(b.Bytes())
-	if err != nil {
-		return errors.Wrapf(err, "failed to run '%s %s' command in directory '%s', output: '%s'",
-			name, strings.Join(args, " "), dir, output)
+	if c.Quiet {
+		return true
 	}
-	return err
+	return false
 }
 
-// RunCommandVerbose runs the given command and logs the output
-func RunCommandVerbose(dir string, name string, args ...string) error {
-	os.Setenv("PATH", PathWithBinary())
-	e := exec.Command(name, args...)
-	if dir != "" {
-		e.Dir = dir
+// Run Execute the command and block waiting for return values
+func (c *Command) Run() (string, error) {
+	os.Setenv("PATH", PathWithBinary(c.Dir))
+	var r string
+	var e error
+
+	f := func() error {
+		r, e = c.run()
+		c.attempts++
+		if e != nil {
+			c.Errors = append(c.Errors, e)
+			return e
+		}
+		return nil
 	}
-	var b bytes.Buffer
-	e.Stdout = &b
-	e.Stderr = &b
-	err := e.Run()
-	output := string(b.Bytes())
+
+	c.ExponentialBackOff = backoff.NewExponentialBackOff()
+	if c.Timeout == 0 {
+		c.Timeout = 3 * time.Minute
+	}
+	c.ExponentialBackOff.MaxElapsedTime = c.Timeout
+	c.ExponentialBackOff.Reset()
+	err := backoff.Retry(f, c.ExponentialBackOff)
```

Comment:
Is this thread safe? I thinking to a case when a command is still in execution but the retry period is reached. Is going to schedule a second execution. I think this could happen at the beginning of the exponential period.

Reply:
I think so...... But I don't have a test for that. Let me add one.